### PR TITLE
new feature: keep only the last N log files in rotation

### PR DESCRIPTION
**Pull request by ozgunozerk**
## Motivation

`tracing-appender` does not have `Rotation` based on size yet.
Also, it doesn't have the feature of **keeping the most recent `N` log files**

I believe the second feature is more easy to implement, and also will partially solve the `Rotation` based on size problem. Because people may choose `hourly` or `daily` rotation based on their needs, and put an extra boundary of `keep the last 5 files` for example. Of course it won't handle all the edge cases for `Rotation` based on size. But it will cover most of the scenarios. And also, it is a good feature to have on its own :)

## Solution

Introduce another field called `keep_last: Option<usize>` to the `Inner` of `RollingFileAppender` struct.
I managed to did not touch any of the existing functions, so it **WON'T BE A BREAKING CHANGE**. Yay :)

The solution is, whenever the rotation should happen, the `refresh_writer()` is called. So I embed the following logic into that function:

1- check the log folder and detect the log files
2- if there are more log files than the `last_keep` amount
3- store the filenames in a vector, and sort them by their dates (dates are already present in the filename)
4- keep deleting the oldest ones, till we have desired amount of log files in the log folder

P.S. this PR was opened before, but got closed since I had to `sync` the fork, so I re-opened it

### CODE REVIEW

**Comment by IniterWorker** [comment]
I think we should consider a more generic way from a library perspective point of view that can solve #1940 and/or provide a customizable abstraction.

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:161]: I think this makes more sense to add this method to the [`Builder` type](https://github.com/tokio-rs/tracing/blob/a77be16e395893a82474688e2f3f54522937e64c/tracing-appender/src/rolling/builder.rs#L30), rather than on the `RollingAppender` itself. In this case, it can be configured on the `RollingAppender` type directly, but I think it's a bit confusing for users to call some configuration methods on the `Builder` type and other methods on the `RollingAppender` after it's built. What do you think?

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:618]: This syntax doesn't compile on our current MSRV for `tracing-appender` (Rust 1.53.0): https://github.com/tokio-rs/tracing/actions/runs/2905294358/jobs/4624104893

Until the MSRV is new enough to support this syntax, we'll have to write this instead:
```suggestion
                                    Err(error) => eprintln!("error parsing the date: {}", error),
```

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:631]: simiilarly, for now, this needs to be 
```suggestion
                            eprintln!("Failed to remove old log file: {}", error);
```

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:637]: and:
```suggestion
                eprintln!("Error reading the log directory/files: {}", error)
```

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:629]: it looks like we're currently determining the age of a log file by reading the filename and attempting to parse the date from the filename, which feels a bit complex and potentially flaky. Instead, I think we could just use the file's creation timestamp, as returned by [`fs::Metadata::created`](https://doc.rust-lang.org/stable/std/fs/struct.Metadata.html#method.created).

Something like this could potentially be a lot simpler:
```suggestion
            Ok(mut files) => {
                if files.len() >= keep_last {
                    // sort the files by their creation timestamps.
                    files.sort_by_key(|file| {
                        file.metadata()
                            .and_then(|metadata| metadata.created().map(Some))
                            .unwrap_or_else(|error| {
                                eprintln!(
                                    "Unable to get file creation time for {}: {}",
                                    file.path().display(),
                                    error
                                );
                                // if we couldn't determine the file's creation
                                // timestamp for whatever reason, it should
                                // compare as "less than" other files, so that
                                // it's not deleted.
                                None
                            })
                    });

                    // delete files, so that (n-1) files remain, because we will create another log file
                    for file in &files[..files.len() - (keep_last - 1)] {
```

File: tracing-appender/src/rolling.rs
Comment (hawkw) [reply tracing-appender/src/rolling.rs:631]: also, we might want to include the path of the file we couldn't delete here?

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:640]: This is a style nit, not a hard blocker, but this function contains a lot of very deeply indented code, which i find kind of hard to read. I think this function would be much more readable with some judicious use of early returns instead. Here's what I might recommend (with my other suggestions applied as well):
```suggestion
        let files = fs::read_dir(&self.log_directory).map(|dir| {
            dir.filter_map(Result::ok)
                .filter(|entry| {
                    entry.file_name().to_string_lossy().contains(
                        &self
                            .log_filename_prefix
                            .clone()
                            .expect("prefix is always present"),
                    )
                })
                .collect::<Vec<_>>()
        });

        let mut files = match files {
            Ok(files) => files,
            Err(error) => {
                eprintln!("Error reading the log directory/files: {}", error);
                return;
            }
        };

        if files.len() < keep_last {
            return;
        }

        // sort the files by their creation timestamps.
        files.sort_by_key(|file| {
            file.metadata()
                .and_then(|metadata| metadata.created().map(Some))
                .unwrap_or_else(|error| {
                    eprintln!(
                        "Unable to get file creation time for {}: {}",
                        file.path().display(),
                        error
                    );
                    // if we couldn't determine the file's creation timestamp
                    // for whatever reason, it should compare as "less than"
                    // other files, so that it's not deleted.
                    None
                })
        });

        // delete files, so that (n-1) files remain, because we will create another log file
        for file in &files[..files.len() - (keep_last - 1)] {
            if let Err(error) = fs::remove_file(file.path()) {
                eprintln!(
                    "Failed to remove old log file {}: {}",
                    file.path().display(),
                    error
                );
            }
        }
```

File: tracing-appender/Cargo.toml
Comment (hawkw) [inline tracing-appender/Cargo.toml:25]: if you take my suggestion about using the file's creation timestamp rather than parsing the date from the filename, we can remove this:
```suggestion
time = { version = "0.3.2", default-features = false, features = ["formatting"] }
```

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:37]: if you take my suggestion about using the file's creation timestamp rather than parsing the date from the filename, we can remove this:
```suggestion
use time::{format_description, Duration, OffsetDateTime, Time};
```

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:592]: if you take my suggestion about using the file's creation timestamp rather than parsing the date from the filename, we can remove this:
```suggestion
```

**Review by hawkw — CHANGES_REQUESTED** [review]
Thanks for working on this @ozgunozerk, and I'm sorry it took so long to get your PR reviewed! This is definitely something I'd like to add support for, I just haven't had the free time to review `tracing` PRs recently.

I commented on some potential changes to the implementation that I think we might want to make.

Also, we generally prefer to base PRs for changes like this on the `master` branch, rather than `v0.1.x`, when the changed code is the same between the two branches. That makes it easier to ensure changes made to the release version won't be "lost" when the `tracing` 0.2 ecosystem is released. Would you mind changing this PR to be based on the `master` branch? Once it merges, a maintainer will handle backporting the change to `v0.1.x` so that it can be released.

@ozgunozerk, if you no longer have the time to continue working on this change, I'd be happy to make the suggested changes to get this over the finish line, and merge it (with credit to you, of course). Please let me know if you'd like me to do that!

Thanks again for the PR!

**Comment by ozgunozerk** [comment]
> @ozgunozerk, if you no longer have the time to continue working on this change, I'd be happy to make the suggested changes to get this over the finish line, and merge it (with credit to you, of course). Please let me know if you'd like me to do that!

@hawkw I'm still willing to spend effort on this, thank you very much for the kind offer :)
I'll start on fixes/suggestions today!

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:629]: did not know about metadata... Great suggestion thanks!

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs]:
```
// if we couldn't determine the file's creation timestamp
// for whatever reason, it should compare as "less than"
// other files, so that it's not deleted.
```

then, won't we delete other log files, which should not be deleted, because they are the recent ones?
I get the idea of your suggestion and I agree, but I think it may create another problem

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:161]: I agree, however in this case, back-compatibility will be a problem, since there will be an explicit new field called `keep_last`, and the users who want to upgrade to new version should specify `None` for the last argument (`keep_last`).
Decision is yours :)

**Comment by ozgunozerk** [comment]
@hawkw when I change the base to master, there are approx. 540 commits inherited because I originally based my changes on v0.1.x

When the changes are ok, I can create clean PR for `master` branch if you like and close this one

File: tracing-appender/src/rolling.rs
Comment (hawkw) [reply tracing-appender/src/rolling.rs]: hmm, i think maybe the best approach is to just give up on deleting any log fiiles if we can't get the creation timestamp for any of them. there's no correct way to handle that situation without potentially deleting the wrong file...

File: tracing-appender/src/rolling.rs
Comment (hawkw) [reply tracing-appender/src/rolling.rs:161]: I don't think there would be a back-compatibility problem. When a new `Builder` is constructed, the value of `keep_last` would default to `None`, and it would be set if a `Builder::keep_last_n_logs` method is called. Otherwise, it will be `None` if the user doesn't call that method. The primary reason we have a `Builder` type is that it allows us to add as many new parameters as we want without having to change existing function signatures.

File: tracing-appender/src/rolling/builder.rs
Comment (hawkw) [inline tracing-appender/src/rolling/builder.rs:186]: nit: i would probably say "log files" rather than "log entries", since "log entries" seems like it could also refer to an individual log line...
```suggestion
    /// Keep the last `n` log files on disk.
```

File: tracing-appender/src/rolling/builder.rs
Comment (hawkw) [inline tracing-appender/src/rolling/builder.rs:187]: maybe we should also add something like this, just to make sure the behavior is clear:
```suggestion
    ///
    /// When a new log file is created, if there are `n` or more
    /// existing log files in the directory, the oldest will be deleted.
```

File: tracing-appender/src/rolling/builder.rs
Comment (hawkw) [inline tracing-appender/src/rolling/builder.rs:205]: let's move this before the examples, please?

File: tracing-appender/src/rolling/builder.rs
Comment (hawkw) [inline tracing-appender/src/rolling/builder.rs:197]: typo:
```suggestion
    ///     .keep_last_n_logs(5) // only the most recent 5 log files will be kept
```

File: tracing-appender/src/rolling/builder.rs
Comment (hawkw) [inline tracing-appender/src/rolling/builder.rs:190]: this example isn't setting a suffix, let's remove this:
```suggestion
```

File: tracing-appender/src/rolling/builder.rs
Comment (hawkw) [inline tracing-appender/src/rolling/builder.rs:206]: take it or leave it: maybe a clearer name for this would be `max_log_files`? what do you think?

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:614]: i think your comment https://github.com/tokio-rs/tracing/pull/2284#discussion_r977347938 was correct and we probably shouldn't just assume that any log files we can't get the creation time of are "infinitely old",  because then we might end up deleting log files that are actually newer. instead, if we encounter an error getting any file's creation timestamp, this method should probably just print the error and return immediately so that we don't delete anything.

**Review by hawkw — CHANGES_REQUESTED** [review]
Overall, this is looking close to ready. I had a few more minor suggestions, mostly for the documentation.

If you have the time, it would also be great to have a test for this?

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs]: I agree, that would be preferable 👍

File: tracing-appender/src/rolling/builder.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling/builder.rs:206]: this is definitely a better name, taking it :)

Comment (ozgunozerk) [reply tracing-appender/src/rolling/builder.rs:206]: I'm also changing `keep_last` to `max_files`, we can always revert it you think `keep_last` was a better name.

File: tracing-appender/src/rolling.rs
Comment (CBenoit) [inline tracing-appender/src/rolling.rs:603]: I think you could use `map` instead of `and_then` directly

Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:603]: I was thinking of this:
```rust
// sort the files by their creation timestamps.
        files.sort_by_key(|file| {
            match file.metadata().and_then(|metadata| metadata.created()) {
                Ok(val) => val,
                Err(why) => {
                    println!("failed reading the timestamp because: {}", why);
                    // if we couldn't determine the file's creation timestamp
                    // for whatever reason, we should return early
                    return;
                }
            };
        });
```

Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:603]: 1 `and_then()` is beneficial imo, since we have to check both `metadata()` and then `created()`. 
Hahaha I even used the words **and then** myself, what a great language <3

**Comment by ozgunozerk** [comment]
@hawkw 
Will add a test tonight or tomorrow, please let me know if it is good to go otherwise :)

File: tracing-appender/src/rolling.rs
Comment (CBenoit) [reply tracing-appender/src/rolling.rs]: Oh, indeed `file.metadata().and_then(|metadata| metadata.created())` looks good! I overlooked the final output was actually an `Option<SystemTime>` 👍

Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs]: Seems like this wouldn't work, because the `return` is not making the function return, but making the match block return

File: tracing-appender/src/rolling.rs
Comment (CBenoit) [reply tracing-appender/src/rolling.rs:586]: Maybe you can do a `filter_map` operation instead and keep creation times along the files. That way you can remove the `expect` added below and simplify things a bit. What do you think?

Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:586]: Hmm I don't think I follow. 
If you meant using `filter_map()` instead of `filter` ->

How to eliminate the `expect` in the below (line 604) ? 
```rust
file.metadata()
                .and_then(|metadata| metadata.created())
``` this structure will always be a `Result` regardless of we have only valid files with timestamps or not. So I believe, `expect` is necessary. 

If you meant another thing, then probably I did not get it :)

Comment (CBenoit) [reply tracing-appender/src/rolling.rs:586]: You are collecting files in a `Result<Vec<DirEntry>>`. You also extract the creation date once on this occasion. You could thus collect a `Result<Vec<(DirEntry, SystemTime)>>` using `filter_map`, and reuse the values already extracted =)

Something like:
```suggestion
                .filter_map(|entry| {
                    let created = entry
                        .metadata()
                        .and_then(|metadata| metadata.created())
                        .ok()?;
                    Some((entry, created))
                })
```

Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:586]: Wonderful, thanks!

**Comment by hawkw** [comment]
The CI failures are unrelated, it looks like the latest Rust release broke the UI tests for the `#[instrument]` macro's error messages: https://github.com/tokio-rs/tracing/actions/runs/3114487346/jobs/5050405220#step:5:1233

Don't worry about that for now.

**Comment by ozgunozerk** [comment]
I've written a test, that will generate 3 log files,
and set the `max_files` to 2, 
so the first created log file should be deleted. 
@hawkw please check the last commit for the test. I mostly copy pasted the `test_make_writer` to generate this one.

File: tracing-appender/src/rolling/builder.rs
Comment (hawkw) [inline tracing-appender/src/rolling/builder.rs:191]: It occurs to me that we might also want to describe the rules that we use to determine whether or not a file is something we can delete. Maybe something like this:
```suggestion
    /// 
    /// Files are considered candidates for deletion based on the following
    /// criteria:
    ///
    /// * The file must not be a directory or symbolic link.
    /// * If the appender is configured with a [`filename_prefix`], the file
    ///   name must start with that prefix.
    /// * If the appender is configured with a [`filename_suffix`], the file
    ///   name must end with that suffix.
    /// * If the appender has neither a filename prefix nor a suffix, then the
    ///   file name must parse as a valid date based on the appender's date
    ///   format.
    ///
    /// Files matching these criteria may be deleted if the maximum number of
    /// log files in the directory has been reached.
```
this way, users should hopefully know not to store other files created by other programs that _happen_ to match the deletion criteria in the same directory.

**Review by hawkw — CHANGES_REQUESTED** [review]
This is definitely getting close, and thanks for adding the test --- it looks right to me.

I have one remaining concern about this change: it looks like the code for determining whether or not a file in the log file directory is a log file or not works by checking if it starts with the appender's filename prefix, and panicking if `self.log_filename_prefix` is `None`. But, the appender can be configured with a filename prefix, a filename suffix, both, or neither. I don't think that the `max_log_files` setting should panic if the appender was configured without a prefix, since that's a valid configuration some users may be using. We should probably fix that before merging this branch. 

Also, one last small docs change: the `Builder::new()` function's documentation has a list of the builder parameter's default values, here: https://github.com/tokio-rs/tracing/blob/a77be16e395893a82474688e2f3f54522937e64c/tracing-appender/src/rolling/builder.rs#L34-L44
let's add that there's no default max file count.

Besides that, I commented on some style tweaks you may want to make, but none of them are blockers for this change, as long as we take care of the issue with the filename prefix being required.

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:602]: ah, my bad, it looks like my suggested change makes the borrow checker unhappy: https://github.com/tokio-rs/tracing/actions/runs/3114778302/jobs/5050994522
fortunately,  `SystemTime` is `Copy`, so we can just write this:
```suggestion
        files.sort_by_key(|(_, &created_at)| created_at);
```

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:580]: `.log_filename_prefix` is `Option<String>`.
Without `clone()`, the borrow checker is not happy:
```
move occurs because `self.log_filename_prefix` has type `Option<String>`, which does not implement the `Copy` trait
```

Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:580]: I think you are completely right on your first comment. When I wrote the code, I wrote it for my specific needs and forgot to change this part to general purpose. 
Your suggestions sounds pretty reasonable to me, implementing now 💯

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:602]: yeah I already fixed that no worries :)

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs]: Adding the check of parsing the filename as a date required me to add some previous stuff back (adding Date and Format related things). If there is better way doing that, I'm open to suggestions. 

If not, you may want to reintroduce the `parsing` feature again maybe?

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:588]: it's kind of unfortunate that we parse a new format description every time `prune_old_logs` is called, and we also parse them separately in the `join_date` function, here: https://github.com/tokio-rs/tracing/blob/08c44200ae7d966439f673bffc69f549b3dd94d1/tracing-appender/src/rolling.rs#L500-L506

this is potentially inefficient, as it will allocate to store the format description, and also parses the string each time, although this isn't a big deal as these functions aren't called that frequently. additionally, though, it means that the format description strings are defined in two places now, both here in `prune_old_logs` and separately in `join_date`...if we change one of the format descriptions, we have to change the other one as well, or `prune_old_logs` won't work anymore.

what if we did a small refactor to move the format description to a field on `Inner`, and create it only once when constructing a new appender? then, we could just use `self.date_format` in both this method and in `join_date`.

File: tracing-appender/src/rolling.rs
Comment (hawkw) [inline tracing-appender/src/rolling.rs:593]: it might also be worth checking that that the entry's metadata indicates that it's a file using [`Metadata::is_file`](https://doc.rust-lang.org/stable/std/fs/struct.Metadata.html#method.is_file). the appender will never create directories or symlinks, so if the log dir happens to contain a symlink or directory whose name matches our log file format (unlikely, but certainly possible), we shouldn't delete it.

File: tracing-appender/src/rolling/builder.rs
Comment (hawkw) [inline tracing-appender/src/rolling/builder.rs:210]: we probably need to add link references for the `filename_prefix` and `filename_suffix` functions here:
```suggestion
    /// log files in the directory has been reached.
    ///
    /// [`filename_prefix`]: Self::filename_prefix
    /// [`filename_suffix`]: Self::filename_suffix
    ///
    /// # Examples
```
(i meant to include them in my suggested changes but I forgot, sorry)

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:588]: When I first opened the PR 2 months ago, my initial attempt was to add `format` to `Inner`. However it was unsuccessful due to the ***lifetime*** necessity. 

I'll call it a day for now (it's been a long day for me), and I can give this problem another go at the weekend.

Comment (hawkw) [reply tracing-appender/src/rolling.rs:588]: if the format descriptions are always `&'static str`s (which they are), the lifetime of the `FormatItem`s should also be `'static`. if you like, i can open a PR against this branch to make that change?

Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:588]: sure, I don't exactly recall at which step I got stuck 2 months ago. Maybe this time it will be a success. But if you have the time, I won't reject this offer :)

Comment (hawkw) [reply tracing-appender/src/rolling.rs:588]: okay, i've gone ahead and opened https://github.com/ozgunozerk/tracing-keep-last-n-logs/pull/1 with that change --- let me know what you think!

Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:588]: wonderful!

File: tracing-appender/src/rolling.rs
Comment (ozgunozerk) [reply tracing-appender/src/rolling.rs:593]: your PR fixes also this, so resolving this convo :)

**Comment by ozgunozerk** [comment]
Thank you for spending this much effort&time into this @hawkw , 
I still recall your request on basing this PR against `main`. As I said, please let me know when you think this PR is in good shape, so I can close this one and open a new one against `main` branch. 
(if I just change the base target, it becomes more than 530 commits against `main`)

**Comment by hawkw** [comment]
> I still recall your request on basing this PR against `main`. As I said, please let me know when you think this PR is in good shape, so I can close this one and open a new one against `main` branch.

Yup, this looks good to me now, and if you're able to open a new PR against the main branch with these changes, that would make things much more convenient for the maintainers. I don't think there's anything else we need to address before merging this, once a new PR against the main branch is created. Thank you for all your hard work on this!

**Comment by ozgunozerk** [comment]
Please follow from #2323

**Pull request closed without merging**
